### ISSUE TITLE
New or enhanced interface for SolveLinearSystemInAbCategory

### ISSUE THREAD

**Issue by zickgraf** [homalg-project/CAP_project#628]
tl;dr: I am mainly interested on feedback regarding the two suggestions marked by `1.` :-)

I am still working on speeding up the computation of lifts (or more general: linear systems) in Freyd categories. In my algorithm I transform lifts in Freyd categories into linear systems in the underlying category as in the discussion at beginning of chapter 6 of the Freyd paper. In this context I have two suggestions for possible improvements/simplifications of `SolveLinearSystemInAbCategory`.

#### First suggestion:
I will now use the notation of the discussion at beginning of chapter 6 of the Freyd paper: We are only interested in the value of X, whereas Y and Z don't matter, they just have to exist. I would like to pass this information down to the underlying category, because in the future we will be able to use this information, see https://github.com/homalg-project/homalg_project/pull/393 (look for `RelativeRightDivide`). I can imagine two different ways for doing that:
1. Use an additional integer `n` to specify that only the solutions of the first `n` variables should be returned.
2. Encode the "interesting" and the "non-intereresting" part in two different arguments similar to `RightDivide` for three arguments.

For the sake of simplicity I would prefer the first solution.

#### Second suggestion:
Constructing the linear system in the underlying categeory (see https://github.com/zickgraf/CAP_project/commit/4d019f17bf9b0df9888213773e6464c87e72c733, although I fear this link will break with the next force push) is a bit cumbersome: I have to create the left and the right coefficients separately (which makes it more difficult to use functional programming for this, which in turn impedes CompilerForCAP), and I have to specify lots of zero morphisms for entries which I am not interested in. These issues could be solved by
1. merging the left and right coefficients into a single list of list of pairs,
2. allowing holes in the lists which implicitely define zero coefficients,
3. introducing the variables as symbols (strings?), so we could actually write something like `[ [ [ rho_A, "X" ], [ "Y", rho_C ] ], [ [ "X", gamma ], [ "Z", rho_B ] ] ]`.

Of course, all points might cause further problems down the road. In particular, the second and third point would require lots of `if` branches in the implementations which are again bad for CompilerForCAP. So for now I would focus on the first point. Are there any applications where having the left and the right coefficients as separate lists is useful?

**Comment by mohamed-barakat** [comment]
In this regard, I would also like to draw your attention to Kamal's [code]([code](https://github.com/homalg-project/CategoryConstructor/blob/9f0e0ee34c77507d1099cb256d102a984b94817f/gap/Tools.gd#L227-L364), which he added to `CategoryConstructor` upon my request. I needed this code to simplify other code in `FunctorCatorgories` and speed up computations, already over finite fields. All of this (and the remaining content of the files `Tools.g?`) should be submitted to CAP at a later point.

**Comment by sebastianpos** [comment]
Your suggestions are both very good.

**First suggestion**
Please go ahead with option 1 if you consider that to be simpler.
But we need to specify a value for the case that all variables are relevant (e.g. `n = 0`, or `n = -1`).
Moreover, if this extra argument is incorporated in `SolveLinearSystemInAbCategory`, then I suggest that 
it should nevertheless be possible to give a function in `AddSolveLinearSystemInAbCategory` that doesn't need that extra argument. You could also think about providing a new primitive for this relative case.

**Second suggestion**
No, the choice of the current interface w.r.t. left/right coefficients was arbitrary. It's good to learn that your proposed handling might be more useful, so please, go ahead.

**Comment by zickgraf** [comment]
Thank you both for the quick reply!

Regarding the first suggestion: I had an additional idea: maybe this could be handled by the compiler (since the compiler knows which variables will be used later in the algorithms and which will simply be ignored). So I will defer the idea until I have a clear idea if this would be possible.

Regarding the second suggestion: Since there is not much code using `SolveLinearSystemInAbCategory` in `CAP_project` I thought the change would be easy, but seeing the code in the link above I'm not so sure anymore. @kamalsaleh @mohamed-barakat Would changing the interface be easy or would this involve a lot of work for you? I think I could put together some convenience functions so you would not have to change your code immediately, but I think we should not keep such a workaround forever.

**Comment by kamalsaleh** [comment]
The code you mention has never been used in any package yet. If  I remember right, it was used only for some experiments and debugging in FunctorCategories. So, changing them is not that urgent, just make sure the package `CategoryConstructors `
still loads without errors :). Later I can adjust the code accourding to the new interface and submit a PR to `CAP_Project`.

**Comment by mohamed-barakat** [comment]
I needed Kamal's code to compute all Endomorphisms of a functor in order to be able to decompose it into indecomposables:

https://github.com/homalg-project/CatReps/blob/master/examples/CategoryOfRepresentations.g#L267

The resulting high-level code is now almost as optimal as the low-level code which was shown to me by Peter Webb.

**Comment by kamalsaleh** [comment]
Here is a little more context from what I remember:
`CatReps` uses the Hom-Structure which is implemented in `FunctorCategories` and the [code](https://github.com/homalg-project/FunctorCategories/blob/f98a92d4b2933976af435f7a2b3a34446eb2a273/gap/HomStructure.gi#L143) there does not use the additional methods in `CategoryConstructor` which are:
- MereExistenceOfUniqueSolutionOfLinearSystemInAbCategory,
- MereExistenceOfUniqueSolutionOfHomogeneousLinearSystemInAbCategory,
- BasisOfSolutionsOfHomogeneousLinearSystemInLinearCategory,
- BasisOfSolutionsOfHomogeneousDoubleLinearSystemInLinearCategory.

The code in the `CategoryConstructor` was meant to solve double linear systems, which can be used to implement Hom-Structure for `FunctorCategories`, but after some experiments we decided not to use them, since creating a double linear system creates too much data that needs to be integrated in any computation. So basically, the mentioned operations in `CategoryConstructor` have not been used yet.

**Comment by mohamed-barakat** [comment]
I see. I forgot these details :)

**Comment by zickgraf** [comment]
Thanks for the information! Then I will proceed with the new interface in the coming days (hopefully) :-)

**Comment by zickgraf** [comment]
I have found a better way to deal with the coefficients: We can interpret the left coefficients and the transpose of the right coefficients as morphisms in the AdditiveClosure of the category. With this, most of the "ugly" iterations over lists are hidden inside AdditiveClosure. Thus, I will not proceed with the new interface and will close this issue.
